### PR TITLE
Refactor handler for workflowIdReusePolicy

### PR DESCRIPTION
**Pull request by stephanos**
## What changed?

- clearer distinction between the behavior for closed and running workflows
- name that isn't particular to workflowIdReusePolicy, but the more general issue of an "ID duplication"

## Why?

Preparing for changes from https://github.com/temporalio/api/pull/359 by pulling this refactor in early, reducing the size of the change later.

## How did you test it?

Existing tests + added a test for Terminate-if-Running earlier in https://github.com/temporalio/temporal/pull/5483.

### CODE REVIEW

File: service/history/api/startworkflow/api.go [hunk 1]
```diff
@@ -332,24 +332,21 @@ func (s *Starter) verifyNamespaceActive(creationParams *creationParams, currentW
 	return nil
 }
 
-// applyWorkflowIDReusePolicy applies the workflow ID reuse policy in case a workflow start requests fails with a
-// duplicate execution.
-// At the time of this writing, the only possible action here is to terminate the current execution in case the start
-// request's ID reuse policy is TERMINATE_IF_RUNNING.
```
Comment (stephanos) [inline service/history/api/startworkflow/api.go:338]: I don;t think this comments needs to talk about the internals of `resolveDuplicateWorkflowID`

File: service/history/api/startworkflow/api.go [hunk 2]
```diff
@@ -282,23 +282,30 @@ func (s *Starter) handleConflict(
...
 	if err := s.createAsCurrent(ctx, creationParams, currentWorkflowConditionFailed); err != nil {
 		return nil, err
 	}
-	return s.generateResponse(creationParams.runID, creationParams.workflowTaskInfo, extractHistoryEvents(creationParams.workflowEventBatches))
+
+	return s.generateResponse(
+		creationParams.runID,
+		creationParams.workflowTaskInfo,
+		extractHistoryEvents(creationParams.workflowEventBatches),
+	)
```
Comment (stephanos) [inline service/history/api/startworkflow/api.go:308]: just a drive-by formatting change: without the newlines I've added, the different lines imply they belong together. but they don't. this way the individual calls are easier to spot.

File: service/history/api/workflow_id_reuse_policy.go [hunk 3]
```diff
@@ -36,71 +36,75 @@ import (
...
+// to the WorkflowIDReusePolicy.
+//
+// An action (ie "mitigate and allow"), an error (ie "deny") or neither (ie "allow") is returned.
+func ResolveDuplicateWorkflowID(
+	workflowID,
+	newRunID,
+	currentRunID string,
+	currentState enumsspb.WorkflowExecutionState,
+	currentStatus enumspb.WorkflowExecutionStatus,
+	currentStartRequestID string,
```
Comment (stephanos) [inline service/history/api/workflow_id_reuse_policy.go:49]: this order makes more sense to me

File: service/history/api/workflow_id_reuse_policy.go [hunk 4]
```diff
@@ -36,71 +36,75 @@ import (
...
-// ApplyWorkflowIDReusePolicy returns updateWorkflowActionFunc
-// for updating the previous execution and an error if the situation is
-// not allowed by the workflowIDReusePolicy.
-// Both result may be nil, if the case is to allow and no update is needed
-// for the previous execution.
-func ApplyWorkflowIDReusePolicy(
-	prevStartRequestID,
-	prevRunID string,
-	prevState enumsspb.WorkflowExecutionState,
-	prevStatus enumspb.WorkflowExecutionStatus,
```
Comment (stephanos) [inline service/history/api/workflow_id_reuse_policy.go:48]: callers are all using "currentXYZ", so the parameters might as well, too

File: service/history/api/workflow_id_reuse_policy.go [hunk 5]
```diff
@@ -36,71 +36,75 @@ import (
...
-		msg := "Workflow execution already finished. WorkflowId: %v, RunId: %v. Workflow Id reuse policy: reject duplicate workflow Id."
-		return nil, generateWorkflowAlreadyStartedError(msg, prevStartRequestID, workflowID, prevRunID)
-	default:
-		return nil, serviceerror.NewInternal(fmt.Sprintf("Failed to process start workflow reuse policy: %v.", wfIDReusePolicy))
-	}
 
-	return nil, nil
+		return UpdateWorkflowWithoutWorkflowTask, workflow.TerminateWorkflow(
+			mutableState,
+			"TerminateIfRunning WorkflowIdReusePolicy",
```
Comment (stephanos) [inline service/history/api/workflow_id_reuse_policy.go:100]: removed second "Policy" at the end

File: service/history/api/workflow_id_reuse_policy.go [hunk 6]
```diff
@@ -36,71 +36,75 @@ import (
...
-				return UpdateWorkflowWithoutWorkflowTask, workflow.TerminateWorkflow(
-					mutableState,
-					"TerminateIfRunning WorkflowIdReusePolicy Policy",
-					payloads.EncodeString(
-						fmt.Sprintf("terminated by new runID: %s", runID),
-					),
-					consts.IdentityHistoryService,
-					false,
-				)
-			}, nil
```
Comment (stephanos) [inline service/history/api/workflow_id_reuse_policy.go:75]: moved into its own function

**Review by yycptt — APPROVED** [review]
